Fix text drawer layering for classical wires

neuralsorcerer opened the pull request:
Changes:
- Ensured `_LayerSpooler` tracks classical writes when sliding nodes from either direction so hybrid operations occupy distinct columns regardless of justification.
- Added text and shared-layering regressions that verify hybrid writers and later conditionals stay ordered for right-justified drawings.

Why?
- Prevent hybrid quantum/classical instructions from overlapping or reordering across all circuit visualizers, even with right justification.
- Fixes #15212

neuralsorcerer committed:
Fix text drawer layering for classical wires

neuralsorcerer committed:
Added a releasenote

alexanderivrii commented on qiskit/visualization/circuit/_utils.py:
```diff
@@ -586,36 +616,31 @@ def is_found_in(self, node, nodes):
-        return not _any_crossover(self.qubits, node, nodes, self.measure_arrows)
+        return not _any_crossover(self.qubits, self.clbits, node, nodes, self.measure_arrows)
 
     def slide_from_left(self, node, index):
         """Insert node into first layer where there is no conflict going l > r"""
-        measure_layer = None
-        if isinstance(node.op, Measure):
-            measure_bit = next(bit for bit in self.measure_map if node.cargs[0] == bit)
-
```
Could you clarify why the function `slide_from_left` needed to be updated?

alexanderivrii commented on qiskit/visualization/circuit/_utils.py:
```diff
@@ -624,36 +649,30 @@ def slide_from_left(self, node, index):
         if not self:
             self.insert(0, [node])
-            inserted = True
+            placed_layer = 0
```
Could you clarify why the function `slide_from_right` needed to be updated?

alexanderivrii commented on qiskit/visualization/circuit/_utils.py:
```diff
@@ -499,42 +499,72 @@ def _sorted_nodes(dag_layer):
+def _get_gate_span(qubits, clbits, node, measure_arrows):
     """Get the list of qubits drawing this gate would cover
     qiskit-terra #2802
     """
```
What does this function return in the presence of both qubits and clbits? (Please also adjust the docstring for this function).

alexanderivrii commented on qiskit/visualization/circuit/_utils.py:
```diff
@@ -499,42 +499,72 @@ def _sorted_nodes(dag_layer):
+            try:
+                wire_indices.append(offset + clbits.index(bit))
+            except ValueError:
+                continue
```
We no longer support `Instruction(...).condition` in Qiskit. There are probably places in the code that still mention it, but it's no longer a thing.

alexanderivrii commented on qiskit/visualization/circuit/_utils.py:
```diff
@@ -499,42 +499,72 @@ def _sorted_nodes(dag_layer):
+        try:
+            wire_indices.append(qubits.index(qreg))
+        except ValueError:
+            continue
```
Why is the `try`-`except` needed here?

alexanderivrii commented on test/python/visualization/test_circuit_drawer.py:
```diff
@@ -209,6 +210,38 @@ def test_wire_order_cregbundle(self):
+        _, _, layers = _utils._get_layered_instructions(circuit)
+        self.assertEqual(
+            [[node.op.name for node in layer] for layer in layers], [["Writer"], ["Reader"]]
+        )
```
We no longer support `Instruction(...).condition`, so I don't think we need this (and the following test).

alexanderivrii commented on test/python/visualization/test_circuit_text_drawer.py:
```diff
@@ -1442,6 +1442,120 @@ def test_text_spacing_2378(self):
         circuit.rz(11111, qr[2])
         self.assertEqual(str(circuit_drawer(circuit, output="text", initial_state=True)), expected)
 
+    def test_text_classical_wires_break_layers(self):
+        """Instructions spanning classical wires do not share layers."""
```
Nice test!

alexanderivrii commented on test/python/visualization/test_circuit_text_drawer.py:
```diff
@@ -1442,6 +1442,120 @@ def test_text_spacing_2378(self):
+        )
+
+    def test_text_classical_wires_break_layers_right_justify(self):
+        """Hybrid classical writes stay separated when right-justified."""
```
Nice test! One small question: what do you mean by "right-justified"?

alexanderivrii commented on test/python/visualization/test_circuit_text_drawer.py:
```diff
@@ -1442,6 +1442,120 @@ def test_text_spacing_2378(self):
+        circuit = QuantumCircuit(2, 1)
+        circuit.append(write, [1], [0])
+        conditional.condition = (circuit.clbits[0], 1)
+        circuit.append(conditional, [0], [])
```
I don't think we need this test because `Instruction.condition` is no longer a thing. However, it might be useful to modify this test to have both instructions involving classical bits and measures.

Update: by experimenting a bit here is indeed an example which was displayed incorrectly before this PR and is displayed correctly now:
```python

    instruction_a = Instruction("NameA", 4, 1, [], label="A")
    instruction_c = Instruction("NameC", 1, 1, [], label="C")
    qc = QuantumCircuit(4, 4)
    qc.append(instruction_a, [0, 1, 2, 3], [0])
    qc.measure(3, 0)
    qc.append(instruction_c, [2], [2])
    print(qc)
```

alexanderivrii commented on qiskit/visualization/circuit/_utils.py:
```diff
@@ -499,42 +499,72 @@ def _sorted_nodes(dag_layer):
+            # the qubit wire without spanning the intermediate wires down to the classical target.
+            # Returning just the directly-touched wires prevents stacked measurements on distinct
+            # classical bits from being forced into separate columns when the registers are bundled.
+            span = [wires[index] for index in sorted(set(wire_indices))]
```
Can you point me to a test (or add a test) that illustrates this complication?

alexanderivrii commented on qiskit/visualization/circuit/_utils.py:
```diff
@@ -499,42 +499,72 @@ def _sorted_nodes(dag_layer):
+                and max_index < offset
+            ):
+                max_index = offset - 1
+            span = wires[min_index : max_index + 1]
```
This is more of a nitpick: the computation seemed a bit more efficient as it was done previously, in particular `min_index` and `max_index` were computed directly, and there was no need to sort sets. Is it possible to follow a similar strategy when working with `clbits` here as well?

alexanderivrii reviewed (commented):
Thank you very much for your contribution, @neuralsorcerer. 

I have verified that this PR fixes the original visualization problem, plus additional problems with circuits that contain both instructions with classical bits and usual measurements.

One important point is that we no longer support setting `condition` on an `Instruction` (there might be some leftover places in the code that mention it, but this capability has been fully replaced by control-flow blocks https://quantum.cloud.ibm.com/docs/en/guides/classical-feedforward-and-control-flow).

Looking at the changes, I agree that we need to update the methods `_get_gate_span` and `_any_crossover` to handle `clbits` as well, but I was not sure why `slide_from_left` or `slide_from_right` also needed to be updated. In fact, I have locally replaced these by their original versions, and your PR still works correctly on all of the examples I tried. (But I am not familiar with the code, and might very well be missing something).

neuralsorcerer committed:
Merge branch 'main' into fix-viz

neuralsorcerer committed:
Fix hybrid circuit layering across drawers and address comments

neuralsorcerer commented:
Thank you @alexanderivrii for your reviews and feedbacks, I've updated with changes. Also, I was doing some experiments and commited some unnecessary changes which I've reverted.

alexanderivrii reviewed (commented):
Sorry for the slow review. I still feel that your changes to `_get_gate_span` can be optimized, in particular avoiding conversions from `sets` to `lists` and vice versa. 

I believe the original code for `_get_gate_span` consists of the following cases:

Case 1: the node is a `ControlFlowOp` but not a `BoxOp`. We want this node to be the only node in the layer, so `_get_gate_span` returns `qubits`, that is the list of all qubits.

Case 2: the node is the standard Z-measurement, and we need to draw arrows from the qubit to the classical bit. In this case `_get_gate_span` returns `qubits[min_index : len(qubits)]`,  that is the list of qubits from the measured qubit to the largest qubit (as we need the space to draw the vertical arrow). 

Case 3: Everything else: the node does not have classical bits, or the node is a standard Z-measurement but we do not need to draw the arrow, or the node is a `BoxOp`. In this case `_get_gate_span` returns `qubits[min_index : max_index + 1]`, spanning the qubits from the smallest qubit used in the node to the largest qubit used in the node. 

The case that the function did not handle correctly is our case: the node is an instruction with classical bits. This should be exactly the same as the case 2: we want to return `qubits[min_index : len(qubits)]`.

So I think we can still keep the function `_get_gate_span` to return only a list of qubits, and replace
```
elif node.cargs and (
    (measure_arrows and isinstance(node.op, Measure)) or getattr(node.op, "condition", None)
):
```
by something like
```
elif node.cargs and (
    (measure_arrows and isinstance(node.op, Measure)) or not isinstance(node.op, Measure))
):
```
What do you think?

neuralsorcerer committed:
Merge branch 'main' into fix-viz

neuralsorcerer committed:
Tighten gate span calculation for hybrid instructions

neuralsorcerer committed:
Fix unused arguments in visualization gate span helpers

neuralsorcerer commented:
Thank you @alexanderivrii for the help, I've been overcomplicating things. I've simplified `_get_gate_span` to reuse qubit slices, covering control-flow, classical-only, and hybrid cases without extra set/list conversions.

alexanderivrii commented on qiskit/visualization/circuit/_utils.py:
```diff
@@ -506,32 +506,34 @@ def _sorted_nodes(dag_layer):
     if isinstance(node.op, ControlFlowOp) and not isinstance(node.op, BoxOp):
         # Because of wrapping boxes for mpl control flow ops, this
         # type of op must be the only op in the layer
         # BoxOps are excepted because they have one block executed unconditionally
-        span = qubits
-    elif node.cargs and (
-        (measure_arrows and isinstance(node.op, Measure)) or getattr(node.op, "condition", None)
+        return qubits
```
Putting this check first makes sense - as it does not need computing `qubit_indices`, thanks!

alexanderivrii commented on qiskit/visualization/circuit/_utils.py:
```diff
@@ -506,32 +506,34 @@ def _sorted_nodes(dag_layer):
+        return qubits
+
+    qubit_index_map = {bit: idx for idx, bit in enumerate(qubits)}
+    qubit_indices = [qubit_index_map[qarg] for qarg in node.qargs if qarg in qubit_index_map]
```
Computing the dict, and computing the indices off of this dict is also a nice performance improvement, thanks! My only comment is that I don't understand under which circumstances `if qarg in qubit_index_map` may be `False`, I tried your code without this `if` and it still seems to work great.

```suggestion
    qubit_index_map = {bit: idx for idx, bit in enumerate(qubits)}
    qubit_indices = [qubit_index_map[qarg] for qarg in node.qargs]
```

alexanderivrii commented on qiskit/visualization/circuit/_utils.py:
```diff
@@ -506,32 +506,34 @@ def _sorted_nodes(dag_layer):
+
+    if not qubit_indices:
+        # Classical-only operations conservatively reserve the circuit width.
+        return qubits
```
I suggest do keep this simplest for the moment. Do we even have operations with classical bits but no quantum bits? We do have the `GlobalPhaseGate` which has neither quantum bits nor classical bits -- and currently does not get drawn, in which case we should indeed return `[]`.

```suggestion
    if not qubit_indices:
        # We should not have operations with only classical wires,
        # and the operations with no qubits shouldn't collide with anything.
        return []
```

alexanderivrii commented on qiskit/visualization/circuit/_utils.py:
```diff
@@ -611,17 +613,17 @@ def slide_from_left(self, node, index):
+                        self.measure_map[bit]
+                        for bit in condition_resources(condition).clbits
+                        if bit in self.measure_map
+                    ),
```
I am now convinced that for your fix to work, we don't need to change any function except for `_get_gate_span`. So I suggest to return this to the original version. (I do think that there is quite a bit of opportunity to optimize and improve this visualization code, but that shouldn't be a part of this PR).

alexanderivrii commented on qiskit/visualization/circuit/_utils.py:
```diff
@@ -611,17 +613,17 @@ def slide_from_left(self, node, index):
-                    except StopIteration:
-                        pass
+                    if carg in self.measure_map and self.measure_map[carg] > index_stop:
+                        index_stop = self.measure_map[carg]
```
Same comment as before: for your fix to work, we don't need to change any function except for _get_gate_span. So I suggest to return this to the original version.

alexanderivrii commented on releasenotes/notes/fix-hybrid-visualization-layering-2c7e3d1a47d3d1d8.yaml:
```diff
@@ -0,0 +1,6 @@
+  - |
+    Fixed all circuit drawers to respect classical write ordering when scheduling hybrid operations,
+    so boxes that touch classical wires no longer overlap when registers are bundled, right justified,
+    or measurement arrows are disabled.
```
A few small suggestions for the release notes. The way they are written right now, they are a bit too technical. Remember they are intended for the users. Also the fixes (or at least the tests) were only for the text circuit drawer.
Something like this:

```suggestion
    Fixed the circuit text drawer so that circuit instructions with classical wires are drawn in separate layers.
```

alexanderivrii reviewed (commented):
Thank you for the progress on this, this is almost ready to go. I had some suggestions about the release notes (they should be less technical as they are intended for the users), and about not changing any code except for `_get_gate_span` (I don't think that these changes affect the fix).

neuralsorcerer committed:
Update suggested changes

neuralsorcerer commented:
Thank you @alexanderivrii for the help. Updated as suggested.

alexanderivrii reviewed (approved):
Thanks for the changes, LGTM!
